packages/number/src/NumberEditor.tsx

```diff
@@ -20,73 +24,150 @@ export interface NumberEditorProps {
   }, [sdkValue]);
 
-  React.useEffect(() => {
-    const stringifiedSdkValue = valueToString(sdkValue);
-    // Update the input value (string) if the SDK value (numeric) changes
-    if (stringifiedSdkValue !== previousValue.current && stringifiedSdkValue !== inputValue) {
-      setInputValue(stringifiedSdkValue);
+  const updateExternalValue = (value: number | undefined) => {
+    if (sdkValue !== value) {
+      setValue(value);
+    }
+  };
+
+  const changeValueByStep = (type: StepChangeType) => {
+    const currentValue = Number.isNaN(+inputValue) ? 0 : +inputValue;
+    let nextValue =
+      type === StepChangeType.Increment ? currentValue + NUMBER_STEP : currentValue - NUMBER_STEP;
+    // Floating point numbers cannot represent all decimals precisely in binary.
+    // This can lead to unexpected results, such as 0.1 + 0.2 = 0.30000000000000004.
+    // See more details: https://floating-point-gui.de/
+    nextValue = +nextValue.toFixed(countDecimals(currentValue));
+
+    setInputValue(valueToString(nextValue));
+    setValue(nextValue);
+  };
+
+  // Keeps focus on the input
+  const handleControlPointerDown: React.PointerEventHandler<HTMLButtonElement> = (event) => {
+    event.preventDefault();
+    inputRef.current?.focus();
+  };
+
+  const handleKeyDown = (event: React.KeyboardEvent<any>) => {
+    const keyToFnMap: {
+      [key: string]: () => void;
+    } = {
+      ArrowUp: () => changeValueByStep(StepChangeType.Increment),
+      ArrowDown: () => changeValueByStep(StepChangeType.Decrement),
+    };
+
+    const fn = keyToFnMap[event.key];
+    if (fn) {
+      event.preventDefault();
+      fn();
+    }
+  };
+
+  const handleInputChange: React.ChangeEventHandler<HTMLInputElement> = (e) => {
+    const value = e.target.value;
+    if (!value) {
+      setInputValue(value);
+      updateExternalValue(undefined);
+      return;
     }
-  }, [inputValue, sdkValue]);
+
+    if (!isNumberInputValueValid(value, field.type)) {
+      return;
+    }
+
+    setInputValue(value);
+
+    const parsedNumber = parseNumber(value, field.type);
+    field.setInvalid(parsedNumber === undefined);
+    if (parsedNumber !== undefined) {
+      updateExternalValue(parsedNumber);
+    }
+  };
 
   return (
-    <div data-test-id="number-editor">
+    <div data-test-id="number-editor" className={styles.container}>
       <TextInput
+        // With type="number" react doesn't call onChange for certain inputs, for example if you type `e`
+        // so we use "text" instead and fully rely on our own validation.
+        // See more details: https://github.com/facebook/react/issues/6556
+        type="text"
         testId="number-editor-input"
-        min={range.min !== undefined ? String(range.min) : ''}
-        max={range.max !== undefined ? String(range.max) : ''}
-        step={field.type === 'Integer' ? '1' : ''}
+        className={styles.input}
+        min={range.min}
+        max={range.max}
+        step={NUMBER_STEP}
```

Comment:
Does this work with text input?

Reply:
Yes, it's actually the same role that input type="number" has: https://www.digitala11y.com/spinbutton-role/

Reply:
I meant the `step` attribute. It doesn't seem to be valid: https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/step

Reply:
step can be removed